### PR TITLE
Added FlowFactory interface and default implementation for coroutines flows

### LINKED ISSUE #7164: Support customizing Coroutine Flows
We should see if there is a way to introduce customization of the Flows we create. Maybe similar to our current `RxObservableFactory`. See https://github.com/realm/realm-java/issues/7163 for context.

### LINKED ISSUE #7165: Add support for Flows with no results when using findFirst
Handling nullability of findFirst() is inconvenient
Problem is that the return of .findFirst() is nullable, and it makes your code ugly. I have something like this in my code now:

```
val myObjectFlow: Flow<MyObject?> = realm.where<MyObject>()
        .contains("Name", name, Case.INSENSITIVE)
        .findFirst()
        ?.toFlow() ?: flowOf(null)
```

I think it would be nice to have the extension function like this:

```
 fun <T : RealmModel> T?.toFlow(): Flow<T?>
```

and in case of this is null, return with flowOf(null)

For context see https://github.com/realm/realm-java/issues/7163

### PR DESCRIPTION
**Pull request by edualonso**
Closes #7164
Closes #7165 

Flows can be tricky to work with. Additionally some users might need to implement a custom emission behaviour - e.g. using a Dropbox Store repository requires disabling the emission of the current value when calling `findAllAsync` upon collection to enable a correct delivery of cached results.

To allow this, we have added the option of configuring a Realm using a `FlowFactory` for coroutines. The default implementation, `RealmFlowFactory`, is addded by default when creating a configuration and works as a port of the `RealmObservableFactory` we already use for supporting custom RxJava flowables.

It was necessary to add Kotlin and Kotlin coroutines dependencies to the main library project. Using `compileOnly` ensures those dependencies will not be present in the delivered artefacts.

Two somewhat important changes have been made to `RealmResults` and `RealmList` in connection with adding support for flows for `DynamicRealm`s. The flow factory is located inside the configuration object, therefore fetching the Realm instance for results and lists is needed to gain access to said factory. However, accessing the Realm instance which owns results and list is forbidden for dynamic Realms from an extension function, which is a restriction that stems from how the type system was built for dynamic and regular Realms. Changing the `getRealm` method in these classes is not feasible due to this very constraint, so we decided to make the `realm` property public and rename it to `baseRealm` (as it already typed `BaseRealm`). We discourage accessing this property directly from any `RealmResults` or `RealmList` instance.

### CODE REVIEW

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt [hunk 1]
```diff
@@ -56,94 +54,25 @@ import kotlinx.coroutines.flow.flowOf
...
-        // Emit current (frozen) value
-        offer(freeze(dynamicObject))
-
-        awaitClose {
-            // Remove listener and cleanup
-            if (!flowRealm.isClosed) {
-                dynamicObject.removeChangeListener(listener)
-                flowRealm.close()
+fun <T : RealmModel> T?.toFlow(): Flow<T?> {
+    // Return flow with object or null flow if this function is called on null
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:58]: Can `this` ever be `null` here?

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt [hunk 2]
```diff
@@ -56,94 +54,25 @@ import kotlinx.coroutines.flow.flowOf
...
+                        ?: throw IllegalStateException("Missing flow factory in Realm configuration.")
+                is DynamicRealm ->
+                    (obj as DynamicRealmObject).let { dynamicRealmObject ->
+                        (realm.configuration.flowFactory?.from(realm, dynamicRealmObject)
+                                ?: throw IllegalStateException("Missing flow factory in Realm configuration.")) as Flow<T?>
+                    }
+                else -> throw UnsupportedOperationException("${realm.javaClass} is not supported as a candidate for 'toFlow'. Only subclasses of RealmModel/RealmObject can be used.")
             }
+        } else {
+            return flowOf(this)
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:75]: Maybe add a comment about unmanaged objects only being emitted once?

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmResultsExtensions.kt [hunk 3]
```diff
@@ -63,38 +57,17 @@ import kotlinx.coroutines.flow.flowOf
...
+    return when (val realmInstance = baseRealm) {
+        is Realm -> realmInstance.configuration.flowFactory?.from(baseRealm as Realm, this)
+                ?: throw IllegalStateException("Missing flow factory in Realm configuration.")
+        is DynamicRealm -> realmInstance.configuration.flowFactory?.from(baseRealm as DynamicRealm, this)
+                ?: throw IllegalStateException("Missing flow factory in Realm configuration.")
+        else -> throw IllegalStateException("Wrong type of Realm.")
     }
 }
+
+// TODO figure out if we want to do this as a separate method or merge both in one that delivers changesets and results
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmResultsExtensions.kt:70]: After thinking a bit about this,  I think we should introduce this method as well. IMO the one just returning the `RealmResults` are far easier to consume, this one is slightly annoying as you have to unpack things everytime, but it still has a lot of uses.

File: realm/realm-library/build.gradle [hunk 4]
```diff
@@ -208,8 +208,14 @@ repositories {
 
 dependencies {
 
+    // Hidden from deliverable
```
Comment (cmelchior) [inline realm/realm-library/build.gradle:211]: I would probably just delete this comment.

File: realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java [hunk 5]
```diff
@@ -796,6 +801,68 @@ public void rxFactory() {
         assertFalse(configuration2.getRxFactory() == dummyFactory);
     }
 
+    @Test
+    public void coroutinesFactory_defaultNotNull() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java:805]: Missing test for what happens if you try to assign `null` to the factory. I assume we want to throw an IllegalArgumentException?

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt [hunk 6]
```diff
@@ -424,4 +431,57 @@ class SyncConfigurationTests {
...
+
+    @Test
+    fun coroutinesFactory_defaultNonNull() {
+        val configuration = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
+                .build()
+        assertNotNull(configuration.flowFactory)
+    }
+
+    @Test
+    fun coroutinesFactory() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt:443]: Same, missing test for what happens if you assign null

File: realm/realm-library/src/main/java/io/realm/OrderedRealmCollectionImpl.java [hunk 7]
```diff
@@ -25,7 +25,13 @@
     private static final String NOT_SUPPORTED_MESSAGE = "This method is not supported by 'RealmResults' or" +
             " 'OrderedRealmCollectionSnapshot'.";
 
-    final BaseRealm realm;
+    /**
+     * The {@link BaseRealm} instance in which this collection resides.
+     * <p>
+     * Warning: we discourage using this property directly from your code.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/OrderedRealmCollectionImpl.java:31]:
```suggestion
     * Warning: This field is only exposed for internal usage, and should not be used.
```

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 8]
```diff
@@ -267,6 +272,18 @@ public RxObservableFactory getRxFactory() {
         return rxObservableFactory;
     }
 
+    /**
+     * FIXME
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:276]: Missing Javadoc

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 9]
```diff
@@ -267,6 +272,18 @@ public RxObservableFactory getRxFactory() {
...
+     * @return
+     */
+    @Nullable
+    public FlowFactory getFlowFactory() {
+        if (flowFactory == null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:281]: We should guard against this in the builder so it can never be null here.

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 10]
```diff
@@ -73,7 +73,14 @@
 
     // Always null if RealmList is unmanaged, always non-null if managed.
     private final ManagedListOperator<E> osListOperator;
-    protected final BaseRealm realm;
+
+    /**
+     * The {@link BaseRealm} instance in which this list resides.
+     * <p>
+     * Warning: we discourage using this property directly from your code.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmList.java:80]:
```suggestion
     * Warning: This field is only exposed for internal usage, and should not be used.
```

File: realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+import io.realm.RealmObject;
+import io.realm.RealmResults;
+import io.realm.annotations.Beta;
+import kotlinx.coroutines.flow.Flow;
+
+/**
+ * Factory interface for creating Kotlin {@link Flow}s for Realm classes.
+ */
+@Beta
+public interface FlowFactory {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java:35]: I think we might want to add the `changeSetsFrom` methods already now to this interface. Even if we don't expose them in this PR, it will prevent a later breaking change?

But then again, the interface is marked Beta.

Thoughts?

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+import io.realm.*
+import io.realm.annotations.Beta
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.flowOf
+
+/**
+ * Flow factory used by Realm by default.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt:27]: Flow 
```suggestion
 * Factory class used to create coroutine Flows. This class is used by default unless overridden in {@link RealmConfiguration.Builder#flowFactory(FlowFactory)}.
```

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt:27]: What actually happens in our Javadoc? Is this class even listed there? If not, that is a problem, that might prevent us from implementing this in Kotlin. 🤔

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java [hunk 13]
```diff
@@ -793,6 +799,17 @@ public Builder rxFactory(RxObservableFactory factory) {
...
         }
 
+        /**
+         * Sets the {@link FlowFactory} used to create coroutines Flows from Realm objects.
+         * The default factory is {@link RealmFlowFactory}.
+         *
+         * @param factory factory to use.
+         */
+        public Builder flowFactory(FlowFactory factory) {
+            flowFactory = factory;
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:809]: We should guard against null here.

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt [hunk 14]
```diff
@@ -56,94 +54,25 @@ import kotlinx.coroutines.flow.flowOf
...
         }
-    }
+    } ?: flowOf(null)
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:77]: We seem to be missing tests for this?

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice 👍  I have a few comments. Some tests seem to be missing, and I would probably recommend exposing the `asChangesetFlows()` as well. Or at least prepare for it, implementation might come in another PR.

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt (hunk 1 above)
Comment (edualonso) [reply realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:58]: Yes, when calling `findFirst().toflow()`. We got feedback from a user and decided to add this here https://github.com/realm/realm-java/issues/7165

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java (hunk 9 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:281]: Is this not the same case as with the Rx factory? It will be null if it isn't present in the classpath. By the way, I have just added a utility method for checking whether Flows are present or not.

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt (hunk 12 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt:27]: I had to add an exception to exclude Kotlin files from being javadoc'ed in the build.gradle script since the mere existence of this class was causing the javadoc task to fail. I think it is totally ignored at the moment. But it is a valid concern. I don't think javadoc can ever document Kotlin classes, it can only be done with dokka, but then I'm not sure how to glue the two outputs together...

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java (hunk 13 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:809]: See https://github.com/realm/realm-java/pull/7191#discussion_r522848928

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt (hunk 14 above)
Comment (edualonso) [reply realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:77]: No, it is here `CoroutinesTests.realmObject_toFlow_nullObjectEmitsNullFlow`.

File: realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java (hunk 11 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java:35]: I will have a look at the current implementation for Rx. I will get back to you once I have a better idea about whether it's doable right away or not.

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java (hunk 9 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:281]: Oh, I see what you mean. Sure, I will mark the parameter as `@Nonnull` (for Kotlin's sake) and throw an exception in case we still receive null in Java.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java (hunk 13 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:809]: Got it - https://github.com/realm/realm-java/pull/7191#discussion_r522857800

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt (hunk 6 above)
Comment (edualonso) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt:443]: This is a Kotlin class, so non-nullability is enforced after adding `@Nonnull` to the Java method.

Comment (cmelchior) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt:443]: No, it doesn't prevent that when calling it from Java. At best you get a warning which can be missed or ignored.

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt (hunk 12 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt]: It seems it might be possible to generate the javadoc using dokka instead: https://blog.jetbrains.com/kotlin/2020/08/dokka-preview-based-on-kotlin-1-4-0-rc/

Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt]: I am unsure about being able to write this class in Java. The file containing the `ProducerScope.awaitClose` extension function doesn't have the `@file:JvmName` that would make it visible from Java, and so far it seems impossible to call it using a `ProducerScope` instance from java 😞

**Review by rorbech — COMMENTED** [review]
Not a thorough review, but had some comments. 

Further, isn't the issue around accessing the `realm` yet another issue of trying to convert a result from the result instead of just constructing the result/flow when needed as suggested in https://github.com/realm/realm-java/issues/7168#issuecomment-721576444 which also would make it easier to provide a reasonable context for execution.

**Comment by edualonso** [comment]
> Not a thorough review, but had some comments.
> 
> Further, isn't the issue around accessing the `realm` yet another issue of trying to convert a result from the result instead of just constructing the result/flow when needed as suggested in [#7168 (comment)](https://github.com/realm/realm-java/issues/7168#issuecomment-721576444) which also would make it easier to provide a reasonable context for execution.

I addressed these concerns with @cmelchior and we decided introducing this solution would make the use of flows an anti-pattern in the SDK. It definitely addresses scenarios where the flow has to be deferred to some point at run time, but it doesn't seem to align to established flow patterns.

We agreed on that we could write a "non-official" gist where we could showcase it, but it seems it is not a candidate, at least for the Java SDK.

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+        if (realm.isFrozen) {
+            return flowOf(results)
+        }
+
+        val config = realm.configuration
+
+        return callbackFlow {
+            // Do nothing if the results are invalid
+            if (!results.isValid) {
+                return@callbackFlow
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt:110]: Completely forgot about these 🤦

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt (hunk 12 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt]: Will add an internal Kotlin class that will be called from a Java class that implements the flow interface. This way we will not have issues with javadoc.

File: realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java (hunk 11 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/FlowFactory.java:38]: I've added the implementation for all changeset methods and tests for them. I have left all the factory methods markes as "beta".

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt [hunk 16]
```diff
@@ -96,6 +98,18 @@ inline fun <reified T : RealmModel> Realm.createEmbeddedObject(parentObject: Rea
...
 
+/**
+ * Creates a [Flow] for a [Realm]. It should emit the initial state of the Realm when subscribed to and
+ * on each subsequent update of the Realm.
+ *
+ * @return Kotlin [Flow] that emit all updates to the Realm.
+ */
+@Beta
+fun Realm.toflow(): Flow<Realm> {
+    return configuration.flowFactory?.from(this)
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt:109]: Why is there a nullability check here when not for DynamicRealm?

File: realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java [hunk 17]
```diff
@@ -796,6 +801,120 @@ public void rxFactory() {
         assertFalse(configuration2.getRxFactory() == dummyFactory);
     }
 
+    @Test
+    public void rxFactory_nullThrows() {
+        try {
+            configFactory.createConfigurationBuilder()
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java:807]: Normally we would do this to prevent accidental bugs to trigger the wrong failure:

```
private RealmConfiguration.Builder builder =  configFactory.createConfigurationBuilder()
try {
  builder.rxFactory(null)
} catch (IllegalArgumentException e {
  //...
}
```

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Factory class used to create coroutine {@link Flow}s.
+ *
+ * This class is used by default unless overridden in {@link io.realm.RealmConfiguration.Builder#flowFactory(FlowFactory}.
+ */
+@Beta
+public class RealmFlowFactory implements FlowFactory {
+
+    private final InternalFlowFactory factory = new InternalFlowFactory(true);
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.java:41]: The argument about frozen should also be exposed as a public constructor for this class, as people cannot otherwise customize it.

**Review by cmelchior — APPROVED** [review]
Had a few minor comments and we need a changelog entry under Enhancements for 10.1.0, otherwise 🏅 💃 🚀

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.java (hunk 18 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.java:41]: That's true, I missed this one.

**Comment by datikaa** [comment]
Hi! I might be late here, but you should consider handling case when `findAllAsync()` is used with `first()`, which could throw a `ClosedSendChannelException`. It's a know issue, but for now there are some good examples how people deal with this https://github.com/Kotlin/kotlinx.coroutines/issues/974

**Comment by edualonso** [comment]
Hi @datikaa. Do you mean `query.findAllAsync().first()` or `query.findFirstAsync()`. Just to be sure.

**Comment by datikaa** [comment]
> Hi @datikaa. Do you mean `query.findAllAsync().first()` or `query.findFirstAsync()`. Just to be sure.

Sorry that I wasn't clear about the issue. So it can be problematic, when you do something like this:
```kotlin
    suspend fun getFoo(): Foo {
        return Realm.getDefaultInstance().use { realm ->
            realm.where<Foo>().findAllAsync().toFlow().first()
        }
    }
```

In this case, the default implementation of `toFlow()` will firstly `offer(freeze())` but before the `awaitClose { ... }` there could be a callback for the `RealmChangeListener` and `offer(listenerResults.freeze())` will be called, leading to a `ClosedSendChannelException`. I'cant really give you a code that does this 100% of the time, but I'm experiencing it quite often. In the meantime I've done some workaround codes. I don't know if you guys even want to deal with this or wait for the coroutines update, but still I think it's worth to note :)
